src/bluez/adapter/mod.rs

```diff
@@ -11,517 +11,478 @@
+                        let path = args.object;
+
+                        if let Some(device) = args.interfaces.get(BLUEZ_INTERFACE_DEVICE) {
+                            adapter.add_device(&path, device).unwrap();
+                        } else if let Some(service) = args.interfaces.get(BLUEZ_INTERFACE_SERVICE) {
+                            adapter.add_attribute(&path, service).unwrap();
+                        } else if let Some(characteristic) =
+                            args.interfaces.get(BLUEZ_INTERFACE_CHARACTERISTIC)
+                        {
+                            adapter.add_attribute(&path, characteristic).unwrap();
+                        }
 
-    /// Properties of the adapter.
-    pub info: HCIDevInfo,
-}
+                        return true;
+                    },
+                )?,
+            );
+        }
 
-impl Adapter {
-    pub fn from_device_info(di: &HCIDevInfo) -> Adapter {
-        info!("DevInfo: {:?}", di);
-        Adapter {
-            name: String::from(unsafe { CStr::from_ptr(di.name.as_ptr()).to_str().unwrap() }),
-            dev_id: di.dev_id,
-            addr: di.bdaddr,
-            typ: AdapterType::parse((di.type_ & 0x30) >> 4),
-            states: AdapterState::parse(di.flags),
-            info: di.clone(),
+        if let Err(error) = self.proxy().start_discovery() {
+            match error.name() {
+                // Don't error if BlueZ has already started scanning.
+                Some("org.bluez.Error.InProgress") => Ok(()),
+                _ => Err(error)?,
+            }
+        } else {
+            debug!("Starting discovery");
+            Ok(())
         }
     }
 
-    pub fn from_dev_id(ctl: i32, dev_id: u16) -> Result<Adapter> {
-        let mut di = HCIDevInfo::default();
-        di.dev_id = dev_id;
+    fn stop_scan(&self) -> Result<()> {
+        if let Some((_t, token)) = self.match_tokens.remove(&TokenType::DeviceDiscovery) {
+            trace!("Stopping discovery listener");
+            self.listener.lock().remove_match(token)?;
+        }
 
-        unsafe {
-            ioctl::hci_get_dev_info(ctl, &mut di)?;
+        if let Err(error) = self.proxy().stop_discovery() {
+            match error.name() {
+                // Don't error if BlueZ has already stopped scanning.
+                Some("org.bluez.Error.InProgress") => Ok(()),
+                _ => Err(error)?,
+            }
+        } else {
+            debug!("Stopping discovery");
+            Ok(())
         }
+    }
 
-        Ok(Adapter::from_device_info(&di))
+    fn peripherals(&self) -> Vec<Peripheral> {
+        self.manager.peripherals()
     }
 
-    pub fn is_up(&self) -> bool {
-        self.states.contains(&AdapterState::Up)
+    fn peripheral(&self, address: BDAddr) -> Option<Peripheral> {
+        self.manager.peripheral(address)
     }
 
-    pub fn connect(&self) -> Result<ConnectedAdapter> {
-        ConnectedAdapter::new(self)
+    fn active(&self, _enabled: bool) {
```

Comment:
Here and elsewhere, can we get some comments on the methods that need to be filled in? Makes it easier to file help wanted bugs or at least say issues are claimed later, if we know what needs to be filled in.